### PR TITLE
Initial plugin code

### PR DESCRIPTION
**Pull request by mtrojanowski**

### CODE REVIEW

**Review by travisspencer — CHANGES_REQUESTED** [review]
The templates should have the same open source head (that references the Apache license).

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    static final String REQUEST_DATA_IN_SESSION = "sendemailaction:request-data-in-session";
+
+    private final boolean _alwaysSendEmail;
+    private final EmailSender _emailSender;
+    private final Optional<AccountManager> _accountManager;
+    private final SessionManager _sessionManager;
+    private final DateTimeFormatter _formatter;
+    private final boolean _emailShouldIncludeIpAddress;
+
+
```
Comment (travisspencer) [inline src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java]: double new lines. remove 1

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    public SendEmailActionAuthenticationAction(SendEmailActionAuthenticationActionConfig configuration)
+    {
+        // TODO: Readme
```
Comment (travisspencer) [inline src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java]: todo when?

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            Email email = prepareEmail(authenticationAttributes);
+
+            try
+            {
+                _emailSender.sendEmail(recipientMail, email, "sendemailaction/email/message");
+            }
+            catch (RuntimeException exception)
+            {
+                _logger.warn(
+                        "Exception during sending of e-mail informing about a new login to: {}, with the following data. username: {}, IP: {}, User-agent: {}, time of login: {}",
```
Comment (travisspencer) [inline src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java]: very long lines. Wrap at 120

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        Attribute requestDataInSession = _sessionManager.get(REQUEST_DATA_IN_SESSION);
+
+        if (requestDataInSession == null)
+        {
+            return AuthenticationActionResult.pendingResult(prompt());
+        }
+
+        if (shouldSendEmail(authenticationAttributes))
+        {
+            String recipientMail = getRecipientEmail(authenticationAttributes.getSubject());
```
Comment (travisspencer) [inline src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java]: add `@Nullable` on `recipientMail`.

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                        _sessionManager.get(USER_AGENT_ATTRIBUTE).getValue(),
+                        _formatter.format(Instant.ofEpochSecond(authenticationAttributes.getAuthTime())),
+                        exception
+                );
+            }
+        }
+
+        return AuthenticationActionResult.successfulResult(authenticationAttributes);
+    }
+
```
Comment (travisspencer) [inline src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java:125]: add `@Nullable` annotation

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationActionDescriptor.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    @Override
+    public Class<? extends SendEmailActionAuthenticationActionConfig> getConfigurationType()
+    {
+        return SendEmailActionAuthenticationActionConfig.class;
+    }
+
+    @Override
+    public Optional<? extends ManagedObject<SendEmailActionAuthenticationActionConfig>> createManagedObject(SendEmailActionAuthenticationActionConfig configuration)
+    {
+
```
Comment (travisspencer) [inline src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationActionDescriptor.java]: blank line

File: src/test/groovy/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationActionTest.groovy [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+import se.curity.identityserver.sdk.service.EmailSender
+import se.curity.identityserver.sdk.service.SessionManager
+import se.curity.identityserver.sdk.service.authenticationaction.AuthenticatorDescriptor
+import spock.lang.Specification
+
+import static io.curity.identityserver.plugin.SendEmailAction.SendEmailActionAuthenticationAction.CLIENT_IP_ATTRIBUTE
+import static io.curity.identityserver.plugin.SendEmailAction.SendEmailActionAuthenticationAction.REQUEST_DATA_IN_SESSION
+import static io.curity.identityserver.plugin.SendEmailAction.SendEmailActionAuthenticationAction.USER_AGENT_ATTRIBUTE
+
+class SendEmailActionAuthenticationActionTest extends Specification
```
Comment (travisspencer) [inline src/test/groovy/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationActionTest.groovy:36]: very good

File: templates/sendemailaction/email/message.txt.vm [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#*
```
Comment (travisspencer) [inline templates/sendemailaction/email/message.txt.vm:1]: wrong copyright header

File: templates/sendemailaction/email/message.vm [hunk 9]
```diff
@@ -0,0 +1,48 @@
+#*
```
Comment (travisspencer) [inline templates/sendemailaction/email/message.vm:1]: also wrong; use Apache one

**Review by travisspencer — CHANGES_REQUESTED** [review]
Good start. 
Long lines and missing `@Nullable` are my big comments. See other comments inline

File: src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java (hunk 2 above)
Comment (mtrojanowski) [reply src/main/java/io/curity/identityserver/plugin/SendEmailAction/SendEmailActionAuthenticationAction.java]: It's done, just the comment left 🤦

File: templates/sendemailaction/email/message.txt.vm (hunk 8 above)
Comment (mtrojanowski) [reply templates/sendemailaction/email/message.txt.vm:1]: Good catch, thanks! It's now fixed in both.

**Comment by mtrojanowski** [comment]
@travisspencer I applied all the proposed changes.

**Review by travisspencer — APPROVED** [review]

**Pull request merged**